### PR TITLE
Get arbitrary vars

### LINKED ISSUE #20: add getVariables method or similar
the idea is a person should be able to say they want variable X from collection Y and variable Z from the metadata, etc and get a data.table w those variables and any relevant entity IDs. this would help people build the 'pass' app vizs from the site, for ex.

### PR DESCRIPTION
**Pull request by d-callan**
depends on https://github.com/VEuPathDB/veupathUtils/pull/46
depends on https://github.com/microbiomeDB/microbiomeComputations/pull/21

resolves #20 

the goal is to let the mbio main package users ask for arbitrary variables from different collections by name, and get them back in a single data table. to do this is adds two methods, one called `getVariables` and one called `getCollectionVariableNames`. the second can be used to identify potential vars of interest to request of the first.

### CODE REVIEW

File: R/methods-MbioDataset.R [hunk 1]
```diff
@@ -283,4 +283,119 @@ setMethod("getCollection", "MbioDataset", function(object, collectionName = char
...
+#' getMetadataVariableNames(microbiomeData::DiabImmune)
+#' variablesDT <- getVariables(
+#'      microbiomeData::DiabImmune, 
+#'      list("metadata" = c("age_months", "sex"),
+#'           "16S (V4) Genus" = "Bacteroides", 
+#'           "WGS Metagenome enzyme pathway abundance data" = "ANAGLYCOLYSIS-PWY: glycolysis III (from glucose)"
+#'      )
+#' )
+#' @param object A Microbiome Dataset
+#' @param variables The names of the variables to return. this should be a named list
```
Comment (asizemore) [inline R/methods-MbioDataset.R:324]: Nit: Capitalize "This" at start of sentence

File: R/methods-MbioDataset.R [hunk 2]
```diff
@@ -283,4 +283,119 @@ setMethod("getCollection", "MbioDataset", function(object, collectionName = char
...
+                return(data.table::data.table())
+            }
+            dt <- veupathUtils::getCollectionData(collection, presentVars)
+            if (collectionIndex %in% collectionsWithDupsIndexes) {
+                ## rename variables to prepend the collection name
+                names(dt)[names(dt) %in% presentVars] <- paste(collectionName, presentVars)
+            }
+            return(dt)
+        } else {
+            stop(sprintf("Collection '%s' does not contain the following variables: %s", collectionName, paste(variableNames, collapse = ", ")))
```
Comment (asizemore) [inline R/methods-MbioDataset.R:375]: Should have different verbiage if the input was for the 'metadata'?

File: R/methods-MbioDataset.R [hunk 3]
```diff
@@ -283,4 +283,119 @@ setMethod("getCollection", "MbioDataset", function(object, collectionName = char
...
+        stop("variables argument must be a named list")
+    }
+
+    ## identify variables w identical names early
+    flattenedVars <- unlist(variables)
+    dups <- unname(flattenedVars[duplicated(flattenedVars)])
+    collectionsWithDups <- lapply(variables, function(x) {dups %in% x})
+    collectionsWithDupsIndexes <- unname(which(collectionsWithDups == TRUE))
+
+    fetchCollectionVariables <- function(collectionIndex) {
```
Comment (asizemore) [inline R/methods-MbioDataset.R:349]: Just a suggestion. I understand why to have the function defined here, but i personally find it hard to follow when functions are defined in functions. If reasonable, i'd move the def outside of a function. But that is just my preference.

File: R/methods-MbioDataset.R [hunk 4]
```diff
@@ -283,4 +283,119 @@ setMethod("getCollection", "MbioDataset", function(object, collectionName = char
...
+            return(dt)
+        } else {
+            stop(sprintf("Collection '%s' does not contain the following variables: %s", collectionName, paste(variableNames, collapse = ", ")))
+        }
+    }
+
+    ## this kind of assumes that metadata are always on ancestor entities of assays
+    ## this will break for user data, when we get there
+    mergeCols <- getSampleMetadataIdColumns(object)
+    mergeCollectionVariables <- function(x, y) {
```
Comment (asizemore) [inline R/methods-MbioDataset.R:382]: this def could pretty easily be moved outside of the function

File: R/methods-MbioDataset.R [hunk 5]
```diff
@@ -283,4 +283,119 @@ setMethod("getCollection", "MbioDataset", function(object, collectionName = char
...
+#' to the metadata. The returned data.table will contain the 
+#' requested variables as columns. If one of the requested
+#' variables cannot be returned, a warning will be printed.
+#' 
+#' @examples
+#' getCollectionVariableNames(microbiomeData::DiabImmune, "16S (V4) Genus")
+#' getMetadataVariableNames(microbiomeData::DiabImmune)
+#' variablesDT <- getVariables(
+#'      microbiomeData::DiabImmune, 
+#'      list("metadata" = c("age_months", "sex"),
```
[the remark this answers is not in the document]
Comment (d-callan) [reply R/methods-MbioDataset.R:318]: maybe, and i almost did that.. but then i noticed that we have duplicates for other reasons too. like the pathway abundances and coverage collections have all the same vars for ex..

File: R/methods-MbioDataset.R (hunk 1 above)
Comment (d-callan) [reply R/methods-MbioDataset.R:324]: never!

File: R/methods-MbioDataset.R (hunk 3 above)
Comment (d-callan) [reply R/methods-MbioDataset.R:349]: i tend to like them nested if i kind of actively dont want someone trying to reuse it for random things.. if i move it ill have to generalize it in a way that it could actually be properly repurpose-able. im not sure how useful or worth the time that is..

File: R/methods-MbioDataset.R [hunk 6]
```diff
@@ -283,4 +283,119 @@ setMethod("getCollection", "MbioDataset", function(object, collectionName = char
...
+setMethod(collectionVarNamesGeneric, "MbioDataset", function(object, collectionName) {
+    return(veupathUtils::getCollectionVariableNames(getCollection(object, collectionName)))
+})
+
+#' Get Microbiome Dataset Variables
+#' 
+#' Get the variables in the Microbiome Dataset by their names.
+#' The requested variables could belong to any collection or
+#' to the metadata. The returned data.table will contain the 
+#' requested variables as columns. If one of the requested
```
[the remark this answers is not in the document]
Comment (d-callan) [reply R/methods-MbioDataset.R:310]: yes

File: R/methods-MbioDataset.R (hunk 2 above)
Comment (d-callan) [reply R/methods-MbioDataset.R:375]: metadata was returned before now..
